File: 第06章-控制理论基础/04-案例/matlab/step_and_poles.m
thisDir = fileparts(mfilename('fullpath'));
root = fileparts(fileparts(fileparts(thisDir)));
addpath(fullfile(root, 'common', 'matlab'));

dt = 0.001; T = 5; n = round(T/dt); t = linspace(0,T,n+1);
wn = 4; zetas = [0.3, 0.7, 1.2]; styles = {'-', '--', ':'};
figure; hold on;
for i = 1:3
    zeta = zetas(i); y = 0; yd = 0; ys = zeros(1,n+1);
    for k = 1:n
        ydd = wn^2*(1-y) - 2*zeta*wn*yd;
        yd = yd + dt*ydd; y = y + dt*yd; ys(k+1) = y;
    end
    plot(t, ys, ['k' styles{i}]);
end
grid on; xlabel('t (s)'); ylabel('y'); legend('0.3','0.7','1.2');
title('Second-order step response');

for vx = [10 20 30]
    A = lateral_AB(vx);
    disp(vx); disp(eig(A).');
end

File: 第06章-控制理论基础/04-案例/python/step_and_poles.py
from __future__ import annotations

import sys
from pathlib import Path

import matplotlib.pyplot as plt
import numpy as np

ROOT = Path(__file__).resolve().parents[3]
sys.path.insert(0, str(ROOT / "common" / "python"))
from lateral_2dof import lateral_AB  # noqa: E402


def step_second_order(zeta: float, wn: float = 4.0, T: float = 5.0, dt: float = 0.001):
    n = int(T / dt)
    y = 0.0
    yd = 0.0
    ys = np.zeros(n + 1)
    for k in range(n):
        ydd = wn**2 * (1.0 - y) - 2 * zeta * wn * yd
        yd += dt * ydd
        y += dt * yd
        ys[k + 1] = y
    t = np.linspace(0.0, T, n + 1)
    return t, ys


def main() -> None:
    fig, ax = plt.subplots(figsize=(7, 4))
    for zeta, ls in [(0.3, "-"), (0.7, "--"), (1.2, ":")]:
        t, y = step_second_order(zeta)
        ax.plot(t, y, "k" + ls, label=f"zeta={zeta}")
    ax.set_xlabel("t (s)")
    ax.set_ylabel("y")
    ax.grid(True)
    ax.legend()
    ax.set_title("Second-order step response")
    fig.tight_layout()
    out = Path(__file__).with_name("step_response.png")
    fig.savefig(out, dpi=120)
    print(f"saved {out}")

    for vx in (10.0, 20.0, 30.0):
        A, _ = lateral_AB(vx)
        eigs = np.linalg.eigvals(A)
        print(f"vx={vx:.0f}  poles = {eigs}")
    plt.show()


if __name__ == "__main__":
    main()
